### PR TITLE
Authorization Logic :: Type Declarations

### PR DESCRIPTION
**Pull request by aferr**
This implements type declarations for authorization logic. Type declarations are needed to support both:
https://b.corp.google.com/issues/219015959
and
https://b.corp.google.com/issues/219015796
for transparent release. The types supported are symbols, numbers, and principals. The syntax is directly borrowed from Souffle for now. An extra space is needed between type declarations and parameter names, but there is https://github.com/google-research/raksha/issues/375 to track fixing this eventually.

### CODE REVIEW

File: rust/tools/authorization-logic/src/souffle/souffle_emitter.rs [hunk 1]
```diff
@@ -15,116 +15,81 @@
  */
 
 use crate::{ast::*, souffle::datalog_ir::*};
-use std::collections::HashSet;
 
-pub struct SouffleEmitter {
-    // This is a set of predicates that need to be declared. This set is
-    // populated by side effect while traversing the program body.
-    decls: HashSet<AstPredicate>,
```
Comment (aferr) [inline rust/tools/authorization-logic/src/souffle/souffle_emitter.rs:23]: Declarations used to be generated by keeping around a set of all the predicates that were encountered and then translating every parameter into a symbol.  This is removed now that there are type declarations.

**Comment by bgogul** [comment]
> This implements type declarations for authorization logic. Type declarations are needed to support both: https://b.corp.google.com/issues/219015959 and https://b.corp.google.com/issues/219015796 for transparent release. 

Links to internal bugs don't make sense in open source. When you get a chance add a brief description of what this PR is about. I have some context and will review it in any case.

File: rust/tools/authorization-logic/src/souffle/lowering_ast_datalog.rs [hunk 2]
```diff
@@ -349,7 +492,15 @@ impl LoweringToDatalogPass {
         };
         dlir_assertions.append(&mut dlir_queries);
         dlir_assertions.push(dummy_assertion);
+
+        // type declarations
```
Comment (bgogul) [inline rust/tools/authorization-logic/src/souffle/lowering_ast_datalog.rs:496]: Replicate the comment about how this should be called after body is emitted?

File: rust/tools/authorization-logic/src/souffle/souffle_emitter.rs (hunk 1 above)
Comment (bgogul) [reply rust/tools/authorization-logic/src/souffle/souffle_emitter.rs:23]: Thanks for the heads up. In github, you can hide whitespace during review if needed.

**Review by bgogul — COMMENTED** [review]
Mostly clarification questions and some minor suggestions. 

Haven't reviewed rust code for style issues.

File: rust/tools/authorization-logic/src/parsing/astconstructionvisitor.rs [hunk 3]
```diff
@@ -226,8 +228,57 @@ fn construct_import(ctx: &ImportAssertionContext) -> AstImport {
...
+            AstType::SymbolType,
+        AuthLogicTypeContextAll::PrincipalTypeContext(ctx_prime) =>
+            AstType::PrincipalType,
+        _ => {
+            panic!("construct_type tried to build error");
+        }
+    }
+}
+
+fn construct_type_declaration(ctx: &TypeDeclarationContext) -> AstTypeDeclaration {
```
[the remark this answers is not in the document]
Comment (aferr) [reply rust/tools/authorization-logic/src/parsing/astconstructionvisitor.rs:245]: `astconstructionvisitor.rs` is the interface between the code that we have written and the parser that is generated by Antlr. This type is defined in code generated by antlr that is produced in `AUTH_LOGIC/src/parsing/antlr_gen`. 

(As you may have noticed, `astconstructionvisitor.rs` is technically not a visitor -- in general ANTLR backends can generate visitors, but the visitor generation for rust is buggy, so we can't use it in this rust code.)

File: rust/tools/authorization-logic/src/souffle/lowering_ast_datalog.rs (hunk 2 above)
Comment (aferr) [reply rust/tools/authorization-logic/src/souffle/lowering_ast_datalog.rs]: ok

File: rust/tools/authorization-logic/test_inputs/conditions [hunk 4]
```diff
@@ -1,3 +1,8 @@
+.decl fact1(x : symbol)
+.decl fact2(x : symbol)
+.decl cond1(x : symbol)
+.decl cond2(x : symbol, y : symbol)
+
```
[the remark this answers is not in the document]
Comment (aferr) [reply rust/tools/authorization-logic/test_inputs/conditions:5]: I think these are going to stay mandatory even for the C++ version as long as we have any notion of types. For languages that have type inference, inference still relies on explicit declarations of types somewhere (for example they might appear in the function signature but not the function body). I don't yet see how type inference would make sense for a language like our authorization logic. So in short, I just think "these are mandatory going forward".

File: rust/tools/authorization-logic/test_inputs/typeDeclarations [hunk 5]
```diff
@@ -0,0 +1,7 @@
+.decl someFact(x : symbol)
+.decl someOtherFact(x : symbol, y : number)
+.decl thirdFact(p : principal)
```
[the remark this answers is not in the document]
Comment (aferr) [reply rust/tools/authorization-logic/test_inputs/typeDeclarations:3]: I think technically since this comment is on the frontend of the auth logic language it does not matter because the backend could still emit `Principal`... but perhaps it's best to keep it stylistically consistent anyway.

File: src/xform_to_datalog/authorization_logic_datalog_facts.cc [hunk 6]
```diff
@@ -23,38 +23,29 @@
...
-      "says_canSay_hasTag",
-      "says_removeTag",
-      "says_canSay_removeTag",
-      "says_may",
-      "says_will",
-      "isAccessPath",
-      "isTag",
-      "isPrincipal"
-  };
+      "says_ownsTag", "says_ownsAccessPath", "says_hasTag",  "says_removeTag",
```
Comment (aferr) [inline src/xform_to_datalog/authorization_logic_datalog_facts.cc:43]: The only change in this file was that the _canSay_ relations are removed from this list because now auth logic can generate those with the types that the C++ side is expecting.

File: src/xform_to_datalog/datalog_facts_test.cc [hunk 7]
```diff
@@ -154,9 +154,12 @@ says_hasTag("particle", "recipe.particle.out", owner, "tag") :- ownsAccessPath(o
 
 
 // Authorization Logic
-.decl grounded_dummy(x0: symbol)
-.decl says_cond1(x0: symbol, x1: symbol)
-.decl says_fact1(x0: symbol, x1: symbol)
+.type SomeType <: symbol
```
Comment (aferr) [inline src/xform_to_datalog/datalog_facts_test.cc:157]: @bgogul Is there any way we can make the tests in `xform_to_datalog` any less brittle? I think I'll have to change all of these tests every time the authorization logic backend is changed.

Comment (aferr) [reply src/xform_to_datalog/datalog_facts_test.cc:157]: There might not be... and even if there is maybe it isn't worth the effort. It might just be that the next PR is the last one where the backed for the rust implementation changes.

File: rust/tools/authorization-logic/Cargo.lock [hunk 8]
```diff
@@ -2,6 +2,15 @@
 # It is not intended for manual editing.
 version = 3
 
+[[package]]
```
Comment (bgogul) [inline rust/tools/authorization-logic/Cargo.lock:5]: I am still not sure why these changes are needed. Did you change your `Cargo.toml` file? If these changes are needed, can you split them off to a separate PR?

File: rust/tools/authorization-logic/src/parsing/AuthLogic.g4 [hunk 9]
```diff
@@ -81,8 +81,19 @@ importAssertion
     : IMPORT principal SAYS ID
     ;
 
+// The name "type" would cause name collisions in the code that antlr generates
+authLogicType
+    : NUMBERTYPE #numberType
+    | PRINCIPALTYPE #principalType
```
Comment (bgogul) [inline rust/tools/authorization-logic/src/parsing/AuthLogic.g4:87]: Any reason to keep this separate from `ID #customType`?

**Review by bgogul — APPROVED** [review]
Looks good except for the following caveats: 
 - Did not look closely at rust code for style. 
 - Split off Cargo.lock changes to a separate PR.

File: rust/tools/authorization-logic/Cargo.lock (hunk 8 above)
Comment (aferr) [reply rust/tools/authorization-logic/Cargo.lock:5]: I didn't change `Cargo.toml`, but some other earlier commit changes it without updating `Cargo.lock`. Sure I can make the update to `Cargo.lock` a separate PR.

File: rust/tools/authorization-logic/src/parsing/AuthLogic.g4 (hunk 9 above)
Comment (aferr) [reply rust/tools/authorization-logic/src/parsing/AuthLogic.g4:87]: Not too many strong reasons at the moment. Some weak reasons though:
 - Extensibility --> extending the grammar of principals is one thing we might naturally do (add and/or on principals for example), and making this a builtin type could help with that. (A bit moot in this case since this is not the implementation we will keep around).
 - Principals can show up in more parts of the grammar than objects of the other types. For example in `canSay` or `canActAs`, so in the implementation when these need to be generated we can call PrincipalType instead of CustomType.

Comment (aferr) [reply rust/tools/authorization-logic/src/parsing/AuthLogic.g4:87]: I'm assuming that the approval means that this is OK to leave for now? If you'd rather get rid of this LMK. We can also just keep this in mind for the C++ impl.

File: rust/tools/authorization-logic/src/parsing/AuthLogic.g4 [hunk 10]
```diff
@@ -81,8 +81,19 @@ importAssertion
     : IMPORT principal SAYS ID
     ;
 
+// The name "type" would cause name collisions in the code that antlr generates
+authLogicType
+    : NUMBERTYPE #numberType
+    | SYMBOLTYPE #symbolType
```
[the remark this answers is not in the document]
Comment (aferr) [reply rust/tools/authorization-logic/src/parsing/AuthLogic.g4]: The tests on both the rust side and the C++ side both pass currently... I think I am a bit confused by your question. What change to the grammar are you anticipating? The `SYMBOLTYPE` is already removed and now just NUMBER, PRINCIPAL, and CUSTOMTYPE are there.

[the remark this answers is not in the document]
Comment (aferr) [reply rust/tools/authorization-logic/src/parsing/AuthLogic.g4]: And `AccessPath` is a `CUSTOMTYPE`.

File: rust/tools/authorization-logic/Cargo.lock (hunk 8 above)
Comment (aferr) [reply rust/tools/authorization-logic/Cargo.lock]: fixed
